### ISSUE TITLE
n_cluster_ versus n_cluster

### ISSUE THREAD

**Issue by sergulaydore** [ahoyosid/ReNA#2]
I am getting some results (for a reduced MNIST dataset) where ``self.n_cluster_`` and ``self.n_cluster`` differ no matter how much I change ``thr`` and ``n_iter``.  Is that something expected? 

https://github.com/ahoyosid/ReNA/blob/28f9fd3916318e3e7e75c7fd526be8a5ab21b893/rena.py#L296

**Comment by bthirion** [comment]
It's hard to say. IIRC, these numbers should be equal at convergence.
Can you provide a dataset on which this occurs concretely ?

**Comment by sergulaydore** [comment]
I used reduced MNIST data set as described [here](http://cognitivemedium.com/rmnist).
I think the number of samples are not enough for convergence.

**Comment by bthirion** [comment]
The code should work with one sample. Maybe something goes wrong if there are many zeros pixels (i.e. pixels identical to their neighbor). 
So you directly apply ReNA to these data and it fails ?
Best,
B

**Comment by bthirion** [comment]
The code should work with one sample. Maybe something goes wrong if there are many zeros pixels (i.e. pixels identical to their neighbor). 
So you directly apply ReNA to these data and it fails ?
Best,
B

**Comment by bthirion** [comment]
Back to that: wan you provide a small script to reproduce the problem ?

**Comment by sergulaydore** [comment]
Sorry for the late response. I wrote [this script](https://gist.github.com/sergulaydore/c3f4fb7e9d35f3c970d232dbe4ce3687) to compare Ward and ReNA. These are my conclusions:

1. If we use full digits data sets from scikit-learn, ReNA works just fine
2. If we use only 5 samples from the digits data set in scikit-learn, ReNA yields a warning about division 0 at line `inv_max = dia_matrix((1. / connectivity_.max(axis=0).toarray()[0], 0),
.`
3. If we use the original MNIST data set from yann.lecun.com, Ward works fine but ReNA gives strange results as shown in the below figure.

Am I missing something here?

![mnist_rena_ward](https://user-images.githubusercontent.com/10912900/51769300-bedd0700-20b0-11e9-88e6-12d81cf67f92.png)

**Comment by ahoyosid** [comment]
Hi @sergulaydore,  thanks for the example.

I notice that for MNIST the weights in the connectivity matrix are rather low or zero, hence the problem with "inv_max". This leads to a percolation problem (a huge connected component).

I propose a hack to make it work, which is just adding a small random noise to avoid the percolation.

I hope it helps.

```
import numpy as np
import matplotlib.pyplot as plt
from nilearn.input_data import NiftiMasker
import nibabel
from os.path import join as pjoin
from torchvision import datasets

from rena import ReNA
from rena import weighted_connectivity_graph
from sklearn.utils import check_random_state

rnd = check_random_state(seed=44)

path = pjoin("YOUR_PATH", "mnist")
data = datasets.MNIST(path, train=True, transform=None, target_transform=None,
                      download=False)

X = data.train_data.data.numpy()[:10, :, :] / 255.

n_samples, n_x, n_y = X.shape

noise = 0.01 * rnd.rand(n_samples, n_x, n_y)
X_data = (X + noise).reshape(n_samples, -1)

n_clusters = 20
mask = np.ones((n_x, n_y, 1)).astype("bool")
affine = np.eye(4)
mask_img = nibabel.Nifti1Image(mask.astype(np.float), affine)
masker = NiftiMasker(mask_img=mask_img, smoothing_fwhm=0,
                     standardize=False).fit()

connectivity_rena = weighted_connectivity_graph(data_matrix=X_data,
                                                masker=masker)

rena = ReNA(scaling=True, n_clusters=n_clusters, masker=masker)

X_red = rena.fit_transform(X_data)
X_hat = rena.inverse_transform(X_red).reshape(n_samples, n_x, n_y)

plt.imshow(X_hat[0], "gray")

plt.figure()
plt.imshow(np.reshape(rena.labels_, (n_x, n_y)),
           interpolation='nearest',  cmap=plt.cm.nipy_spectral)
plt.xticks(())
plt.yticks(())
plt.title('ReNA Labels')

```

**Comment by bthirion** [comment]
Actuallu, if you don't want to introruce too much noise, you maight want to do the following:
```
X += .001 * np.random.randn(*X.shape)
X_ = X / X.std(0)
```
then use X_ to train ReNA

As explained by @ahoyosid  the idea is that very flat regions end up very quickly into large clusters. In the consecutive steps, these large clusters "eat" the others.
renormalizing X into X_ avoids creating these large clusters.

In other words: there is an underlying assumption of reNA that the data are spatially stationary, which is OK in brain imaging, not in MNIST.
